presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -967,6 +1024,17 @@ public static Object invoke(ConnectorSession session, MethodHandle handle, List<
         }
     }
 
+    @VisibleForTesting
+    @NotNull
+    public static Object createFailureFunction(Type type, RuntimeException e)
+    {
+        checkNotNull(e, "Exception is null");
+        String failureInfo = JsonCodec.jsonCodec(FailureInfo.class).toJson(Failures.toFailure(e).toFailureInfo());
+        FunctionCall jsonParse = new FunctionCall(QualifiedName.of("json_parse"), ImmutableList.of(toExpression(Slices.utf8Slice(failureInfo), VARCHAR)));
+        Cast nullCast = new Cast(new NullLiteral(), type.toString());
+        return new FunctionCall(QualifiedName.of("fail"), ImmutableList.of(jsonParse, nullCast));
```

Comment:
I think you could simplify this by making `fail` return `unknown` and then adding the cast to the return value instead of passing in a separate parameter, so it would end up being `cast(fail('json here...') as desired_type)`